### PR TITLE
[READY]Refactors click intercept code

### PR DESCRIPTION
**Pull request by silicons**
Title. Now all datums can be click intercepts, defaulting to not blocking clicks, and you can have multiple intercepts per mob.

### CODE REVIEW

File: code/datums/datum.dm [hunk 1]
```diff
@@ -11,6 +11,28 @@
     var/last_find_references = 0
 #endif
 
+//Return TRUE to prevent clickcode execution.
+/datum/proc/InterceptClickOn(mob/M, atom/target, params)
+	return FALSE
+
+/datum/proc/InterceptDblClickOn(mob/M, atom/target, params)
+	return FALSE
```
Comment (Cyberboss) [inline code/datums/datum.dm]: C O M P O N E N T I Z E

**Review by Cyberboss — DISMISSED** [review]

File: code/datums/datum.dm (hunk 1 above)
Comment (AnturK) [reply code/datums/datum.dm]: This, also most of these are still client persistent so that needs to be supported.

File: code/_onclick/ai.dm [hunk 2]
```diff
@@ -10,8 +10,12 @@
 	Note that AI have no need for the adjacency proc, and so this proc is a lot cleaner.
 */
 /mob/living/silicon/ai/DblClickOn(var/atom/A, params)
-	if(client.click_intercept)
-		if(call(client.click_intercept, "InterceptClickOn")(src, params, A))
+	if(LAZYLEN(click_intercepts))
+		var/block = FALSE
+		for(var/i in click_intercepts)
+			var/datum/D = i
+			block += D.InterceptDblClickOn(src, A, params)
```
Comment (RemieRichards) [inline code/_onclick/ai.dm]: is it really intercepted if every other intercept also happens?

**Comment by deathride58** [comment]
>You can have multiple intercepts per mob

Does this mean we'll eventually be able to phase out the target selection thing in favor of the area you click on a spaceman being the area you target?

**Comment by RemieRichards** [comment]
@deathride58 not what 'click intercept' means.

**Comment by arsserpentarium** [comment]
can somebody write articles on wiki about components and subsystems?

**Comment by arsserpentarium** [comment]
the most fucking part, that it isn't enought to simulate click. Because most object have tons of snowflake user check, if user is advancedtooluser, alive and capable to do something. I've tried to directly call attackby. And failed.

**Comment by optimumtact** [comment]
why does the code look worse

**Comment by AnturK** [comment]
Will reopen when you're done with the other n-staling refactors.

**Pull request closed without merging**